pkg/agent/openflow/pipeline.go

```diff
@@ -2292,26 +2291,25 @@ func (f *featureService) serviceLearnFlow(groupID binding.GroupIDType,
 	svcPort uint16,
 	protocol binding.Protocol,
 	affinityTimeout uint16,
-	nodeLocalExternal bool,
-	svcType v1.ServiceType) binding.Flow {
+	externalAddress bool,
+	nodePortAddress bool) binding.Flow {
 	// Using unique cookie ID here to avoid learned flow cascade deletion.
 	cookieID := f.cookieAllocator.RequestWithObjectID(f.category, uint32(groupID)).Raw()
-	var flowBuilder binding.FlowBuilder
-	if svcType == v1.ServiceTypeNodePort {
-		unionVal := (ToNodePortAddressRegMark.GetValue() << ServiceEPStateField.GetRange().Length()) + EpToLearnRegMark.GetValue()
-		flowBuilder = ServiceLBTable.ofTable.BuildFlow(priorityLow).
-			Cookie(cookieID).
-			MatchProtocol(protocol).
-			MatchRegFieldWithValue(NodePortUnionField, unionVal).
-			MatchDstPort(svcPort, nil)
+	flowBuilder := ServiceLBTable.ofTable.BuildFlow(priorityLow).
+		Cookie(cookieID).
+		MatchProtocol(protocol).
+		MatchDstPort(svcPort, nil)
+
+	// EpToLearnRegMark is required to match the packets that have done Endpoint selection.
+	regMarksToMatch := []*binding.RegMark{EpToLearnRegMark}
+	// If "nodePortAddress" is true, which means that Service type is NodePort, "ToNodePortAddressRegMark" must be used
+	// to match packets. Otherwise, Service IP must be used to match packets.
```

Comment:
`If "nodePortAddress" is true, which means that Service type is NodePort` is too verbose. 
Perhaps just `Match ToNodePortAddressRegMark if it's for NodePort address, otherwise match the provided Service IP.`

Reply:
ditto